pkg/controller/clusterissuers/sync.go

```diff
@@ -20,11 +20,9 @@ func (c *Controller) Sync(ctx context.Context, iss *v1alpha1.ClusterIssuer) (err
 	issuerCopy := iss.DeepCopy()
 	i, err := c.issuerFactory.IssuerFor(issuerCopy)
 
-	if err != nil {
-		return err
+	if err == nil {
+		err = i.Setup(ctx)
```

Comment:
Yes it is intentional. It gives the opportunity to the factory to generate events. For that we need to continue on and not call setup but process the rest of the method. If we stick to the previous behavior some validation would be better placed in setup instead of the factory method.

Reply:
Could you rearrange to something like:

```
func (c *Controller) Sync(ctx context.Context, iss *v1alpha1.Issuer) (err error) {
	issuerCopy := iss.DeepCopy()
	defer func() {
		if saveErr := c.updateIssuerStatus(issuerCopy); saveErr != nil {
			errs := []error{saveErr}
			if err != nil {
				errs = append(errs, err)
			}
			err = errors.NewAggregate(errs)
		}
	}()

	i, err := c.issuerFactory.IssuerFor(issuerCopy)
	if err != nil {
		s := messageErrorInitIssuer + err.Error()
		glog.Info(s)
		c.recorder.Event(issuerCopy, v1.EventTypeWarning, errorInitIssuer, s)
		return err
	}

	err = i.Setup(ctx)
	return err
}
```

The defer doesn't need to happen after the call to IssuerFor with this change, because it will be called regardless whether err == nil. Saves the awkward syntax with err == nil and err != nil